[torchrec] [composable] update ShardedEmbeddingBagCollection to be use registered EBCs with shardedTensors as registered modules (#758)

colin2328 opened the pull request:
Summary:
X-link: https://github.com/pytorch/torchrec/pull/758

This PR fixes a bug in FSDP/DDP, where ShardedTensors are not supported even if passed in as params to ignore. 
this is important for composability because TorchRec named_parameters() will return FQN of shardedTensors (as defined in goals)
It defines device of ShardedTensor to be None when local_tensor() does not exist on rank

update ShardedEmbeddingBagCollection to be composable according to https://docs.google.com/document/d/1TBJSd5zgEg6cRcXv3Okuj7bBkqQwGS2IPh4TLWNNzFI/edit


Differential Revision: D40458625

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -552,11 +552,15 @@ def __init__(
+            self.parameters_to_ignore = set(module._ddp_params_and_buffers_to_ignore)
+        else:
+            self.parameters_to_ignore = set()
+        self._module_parameters = [p for n, p in module.named_parameters() if n not in self.parameters_to_ignore]
```
possibly we just need to move '_build_params_for_reducer()' to here? cc @rohan-varma  and @yhcharles

zhaojuanmao reviewed (commented):
Please add a unit test for the case that the codes are addressing, thanks

@wanchaol would you please help reviewing the shard_tensor part? 

@rohan-varma and @yhcharles would you please help reviewing the DDP part?

wanchaol commented on torch/distributed/_shard/sharded_tensor/_ops/tensor_ops.py:
```diff
@@ -43,7 +43,7 @@ def tensor_device(types, args=(), kwargs=None, pg=None):
         raise TypeError("input needs to be a ShardedTensor")
 
-    return self_st.local_shards()[0].tensor.device
+    return self_st.local_shards()[0].tensor.device if self_st.local_shards() else None
```
This looks ok if no local shards return None, I'm wondering in the case of missing shards, do you want to return None, or if you want to return `torch.cuda.current_device()` instead?

colin2328 replied:
> I'm wondering in the case of missing shards

What do you mean by missing shards?

My assumption is that if local_shards() returns None, then there is not local shards for this ShardedTensor (even though this rank is still aware). IMO NONE is the correct device, but I'm open to other options

wanchaol commented on torch/nn/parallel/distributed.py:
```diff
@@ -646,8 +644,8 @@ def __init__(
         # Check that a module does not have Uninitialized parameters
-        for param in module.parameters():
-            if isinstance(param, torch.nn.parameter.UninitializedParameter):
+        for name, param in module.named_parameters():
```
shall we directly check `self._module_parameters` ?

colin2328 replied:
I think this works

wanchaol reviewed (approved):
sharded_tensor part looks good to me, left a few questions.

zhaojuanmao reviewed (changes requested):
need a unit test

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -646,8 +644,8 @@ def __init__(
-        for param in module.parameters():
-            if isinstance(param, torch.nn.parameter.UninitializedParameter):
+        for name, param in module.named_parameters():
+            if isinstance(param, torch.nn.parameter.UninitializedParameter) and name not in self.parameters_to_ignore:
```
Let's add a unittest where we have a parameter that's ignored and it is unitialized? 

Also, this is technically BC breaking, how do we want to deal with that?

colin2328 replied:
sure, I can add a test showing it breaking with ShardedTensor when ignored  today, and this fixes

re BC breaking -> IMO this is bug fix, not BC breaking functionality.

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -552,11 +552,15 @@ def __init__(
         Joinable.__init__(self)
         self.logger = None
-        if not any((p.requires_grad for p in module.parameters())):
+        if hasattr(module, "_ddp_params_and_buffers_to_ignore"):
```
could you explain the reasoning for this change?

YLGH replied:
this seems like an alternative to https://github.com/pytorch/pytorch/pull/81136, 
there are some sanitization logics (check device etc) that shouldn't apply to sharded tensors. However, even if we set that sharded tensor in params_and_buffers to ignore, sanitization is still run on them, 

this would bypass that

colin2328 replied:
we want to reuse the _module_parameters (where we ignore ignored ones). this is setting these as property of class to be reused

rohan-varma reviewed (commented):
Have a couple of questions, also, both of the changes to DDP are technically BC breaking. What's the proposal to handle that?

rohan-varma reviewed (changes requested):
requesting changes specifically for BC breakage part to be discussed and figured out. Happy to align here.

YLGH commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -630,6 +630,8 @@ def cuda(
         return st_cuda
 
     def to(self, *args, **kwargs) -> ShardedTensor:
+        if not self._local_shards:
```
I think this part could be slightly dangerous/tricky - for example the following:
st = ShardedTensor:
   rank 0 has shard_0
   rank 1 has no shards
 
we call st.to("cuda")

rank 0 will call a ShardedTensor.init_from_local_shards,
rank 1 will return self, 

this will cause rank 0 to block,

I think we may need to do a collective call the the beginning just to check if everyones device are the same as the passed in device. If not the same, then everyone needs to call init_from_local_shards,

cc @wanchaol

wanchaol replied:
I think this might be ok as `init_from_local_shards_and_global_metadata` is not calling collectives. But we need to align the behaviors indeed. `to` generally return a new Tensor, we should return a new ShardedTensor that have empty shards on this rank (with the same global sharded metadata, i.e. call `init_from_local_shards_and_global_metadata`), instead of returning self directly. This makes sure we have a different ShardedTensor object from both empty shards and non-empty shards. @colin2328

colin2328 replied:
@wanchaol , actually, I believe that if the current device and the to device are the same, it is a noop and returns the same tensor

>>> y = x.to("cuda")
>>> z = y.to("cuda")
>>> y.data_ptr() == z.data_ptr()
True

I believe the right thing to do is to return self in this case

wanchaol replied:
Oh that make sense, what I mean is that we should basically check `global_sharded_metadata`, and if every rank have empty shards, we can return self, but if some rank have empty shards while others not, we should return a new ShardedTensor with empty shards on those rank.

colin2328 replied:
ah, I see what you mean.

wanchaol commented on test/distributed/test_c10d_gloo.py:
```diff
@@ -1754,6 +1761,49 @@ def forward(self, x):
 
+    @requires_gloo()
+    @skip_if_lt_x_gpu(2)
+    def ftest_ignored_sharded_tensor(self):
```
nit: remove `f` to trigger CI?

rohan-varma commented on test/distributed/test_c10d_gloo.py:
```diff
@@ -1754,6 +1761,49 @@ def forward(self, x):
+                module=m,
+                params_and_buffers_to_ignore={'st'}
+            )
+            # test to make DDP constructor will not fail when module includes a ShardedTensor when ignored
```
how did DDP fail previously?

colin2328 replied:
it failed because some of the behavior was undefined with shardedTensors (such as the device of a ST when the local_tensor doesnt exist for this rank). Even though this was being passed in as ignored params, DDP would still check m.parameters()

rohan-varma commented on test/distributed/test_c10d_gloo.py:
```diff
@@ -1754,6 +1761,49 @@ def forward(self, x):
+        local_shards = [Shard(torch.randn(5, 10, device=device), local_shard_metadata)]
+        st = init_from_local_shards(local_shards, [10, 10])
+        m = MyModule(st)
+        with _ddp_replicated_tensor(False):
```
are we interested in testing `with _ddp_replicated_tensor(True)`, and if not, why not?

colin2328 replied:
this fails with ShardedTensors (cc @wanchaol ) and my tests don't fix. TBH I don't know how this is controlled, but this is off for TorchRec . @wanchaol  can you comment?

wanchaol replied:
I actually don't know why it was failing when we turn the flag to be True, iirc `ddp_replicated_tensor` was there for some optimizations with regards to replicated tensor, I feel it's ok to skip testing it in this patch.

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -553,11 +553,15 @@ def __init__(
+        else:
+            self.parameters_to_ignore = set()
+        self._module_parameters = [p for n, p in module.named_parameters() if n not in self.parameters_to_ignore]
+        if not any((p.requires_grad for p in self._module_parameters)):
```
I was concerned about later checks assuming that `self._module_parameters` has length at least 1, but if it didn't, then this would throw, so those future checks should be safe. Just confirming this understanding.

rohan-varma reviewed (approved):
LGTM. I think the BC concerns are actually not so dire, because this only changes behavior when parameters are ignored which is a private API at the moment anyways. Thanks for this change!

wanchaol commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -630,7 +630,7 @@ def cuda(
 
     def to(self, *args, **kwargs) -> ShardedTensor:
-        current_device = self._local_shards[0].tensor.device
+        current_device = self._local_shards[0].tensor.device if self._local_shards else torch.cuda.current_device()
```
nit: looks like lint fails, I think it's because `torch.cuda.current_device` returns an int. We should wrap it with `torch.device(torch.cuda.current_device)` to fix the lint.

Also, the ShardedTensor might not always be a "cuda" device as always, so when local shards is empty, we shall check if `self.pg._get_backend_name() == "gloo"`, if so, we mark current_device as cpu, else we use `torch.cuda.current_device()`?

colin2328 committed:
[torchrec] [composable] update ShardedEmbeddingBagCollection to be use registered EBCs with shardedTensors as registered modules (#88026)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/88026

update ShardedEmbeddingBagCollection to be composable according to https://docs.google.com/document/d/1TBJSd5zgEg6cRcXv3Okuj7bBkqQwGS2IPh4TLWNNzFI/edit

this works with DMP

named_parameters() behavior changes -> use include_fused as temporary flag to gate this behavior
note that due to ShardedTensor not supporting grads directly, this won't work for Dense compute kernels when non data parallel. This is not used today, and will add a TODO but is low pri

Test Plan:
test_embeddingbag.py and test_model_parallel

GHA tests will fail bc pytorch PR https://github.com/pytorch/pytorch/pull/88026 is not landed

Differential Revision: D40458625

fbshipit-source-id: 013d9ae761c0548dc48a6c0d1faa47fedefb2f36

colin2328 commented:
@pytorchmergebot

colin2328 commented:
@pytorchmergebot  -g

colin2328 commented:
@pytorchbot merge -g
